[SYCL] Add SYCLPropagateAspectUsage Pass

maksimsab opened the pull request:
This patch introduces a new IR pass from [OptionalDeviceFeatures](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md) doc which propagates aspect usages from declared types to functions which somehow use these types.

maksimsab committed:
[SYCL] Add SYCLPropagateAspectUsage Pass

maksimsab committed:
remove trash output.ll

maksimsab committed:
add support for double

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,282 @@
+  auto TypesWithAspects = GetTypesWithAspectsFromMetadata(M);
+  PrintMap(TypesWithAspects);
+
+  Graph g;
+  for (Function &F : M.functions()) {
+    auto CC = F.getCallingConv();
+    if (CC != CallingConv::SPIR_FUNC && CC != CallingConv::SPIR_KERNEL)
+      continue;
+
+    g.nodes.push_back(Graph::Node(F));
+    g.functionIndex[&F] = g.nodes.size() - 1;
+  }
```
I feel like this graph traversal can be implemented in a bit more simple way, without introducing extra classes/structures:
```c++

using FunctionToAspectsMapTy = DenseMap<Function *, SmallSet<int, 4>>;
FunctionToAspectsMapTy FunctionToAspectsMap;
using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
CallGraphTy CallGraph;
SmallVector<Function *, 16> Kernels;

for (Function &F : M.functions()) {
  if (/* not spir func or spir kernel */)
    continue;
  if (/* is kernel */)
    Kernels.push_back(&F);
  
  for (const auto &I : F.instructions()) {
    using RangeTy = /* ... */;
    RangeTy Range = getAspectsUsedByInstruction(I, TypesWithAspects);
    FunctionToAspectsMap[F].insert(Range.first, Range.second);
    // while we are here, let's update call graph
    if (auto *CI = dyn_cast<CallInst>(I)) {
      if (!CI->isIndirectCall())
        CallGraph[F].insert(CI->getCalledFunction());
    }
  }
}

SmallPtrSet<Function *, 16> Visited;
for (Function &F : Kernels) {
  propagateAspects(F, CallGraph, FunctionsToAspectsMap, Visited);
}

void propagateAspects(Function *F, const CallGraphTy &CallGraph, FunctionToAspectsMapTy &AspectsMap, SmallPtrSet<Function *, 16> Visited) {
  SmallSet<int, 8> LocalAspects;
  for (Function *Callee : CallGraph[F]) {
    if (!Visited.contains(Callee)) {
      Visited.insert(Callee);
      propagateAspectsImpl(Callee, CallGraph, AspectsMap, Visited);
    }
    LocalAspects.insert(AspectsMap[Callee].begin(), AspectsMap[Callee].end());
  }

  // update aspects map to re-use it next time
  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
  
  // set metadata to the function
  // ...
  F->setMetadata(/* ... */);
}
```

AlexeySachkov commented on llvm/test/SYCLPropagateAspectUsage/aggregate_types.ll:
```diff
@@ -0,0 +1,71 @@
+; RUN: opt --PropagateAspectUsage < %s -S
+; | FileCheck %s
```
```suggestion
; RUN: opt --PropagateAspectUsage < %s -S | FileCheck %s
```

AlexeySachkov commented on llvm/test/SYCLPropagateAspectUsage/aggregate_types.ll:
```diff
@@ -0,0 +1,71 @@
+; RUN: opt --PropagateAspectUsage < %s -S
+; | FileCheck %s
+;
+; Test checks that no metadata propagates because MyStruct
```
Description doesn't seem to be correct

AlexeySachkov commented on llvm/test/SYCLPropagateAspectUsage/aggregate_types.ll:
```diff
@@ -0,0 +1,71 @@
+; isn't used inside functions.
+
+; Optional
+%A = type { i32 }
```
`clang` emits struct types with `struct.` or `class.` prefix - this helps to distinguish types from regular variables, I suggest to re-use this approach in tests as well

maksimsab replied:
clang does it. But it doesn't influence pass behavior and the current version is simpler. Also I can see similar approach in other tests. [Example](https://github.com/llvm/llvm-project/blob/main/llvm/test/Analysis/BasicAA/128-bit-ptr.ll)

AlexeySachkov commented on llvm/test/SYCLPropagateAspectUsage/aggregate_types.ll:
```diff
@@ -0,0 +1,71 @@
+; Contains a pointer on %D which contains optional %A
+%F = type { %B, %C*, %D1* }
+
+; CHECK: dso_local spir_kernel void @kernel() !intel_used_aspects !1 {
```
```suggestion
; CHECK: dso_local spir_kernel void @kernelD1() !intel_used_aspects !1 {
```

AlexeySachkov commented on llvm/test/SYCLPropagateAspectUsage/aggregate_types.ll:
```diff
@@ -0,0 +1,71 @@
+  ret void
+}
+
+define dso_local spir_kernel void @kernelD2() {
```
```suggestion
; CHECK: dso_local spir_kernel void @kernelD2() {
define dso_local spir_kernel void @kernelD2() {
```

To check that no metadata was attached to that kernel. Since the pass does propagation of aspects through the whole call stack, I suggest we append a `CHECK` line to _each_ function to check that each of them was handled correctly

AlexeySachkov commented on llvm/test/SYCLPropagateAspectUsage/output.ll.ll:
```diff
@@ -0,0 +1,22 @@
+; ModuleID = 'output.ll'
```
This file should be removed, I suppose. As well as `llvm/test/SYCLPropagateAspectUsage/output.ll`

AlexeySachkov reviewed (changes requested):
First iteration of review

maksimsab committed:
add warning support

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
+void CheckDeclaredAspects(LLVMContext &C, const Function *F,
+                          const Container &Aspects) {
+  MDNode *MDN = F->getMetadata("intel_declared_aspects");
+  if (MDN == nullptr)
```
```suggestion
  if (!MDN)
```

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
+  SmallSet<int, 10> DeclaredAspects;
+  for (unsigned i = 0; i != MDN->getNumOperands(); ++i) {
+    const ConstantAsMetadata *CM =
+        dyn_cast<const ConstantAsMetadata>(MDN->getOperand(i));
```
```suggestion
        cast<const ConstantAsMetadata>(MDN->getOperand(i));
```

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
+    const ConstantAsMetadata *CM =
+        dyn_cast<const ConstantAsMetadata>(MDN->getOperand(i));
+    DeclaredAspects.insert(
+        dyn_cast<const ConstantInt>(CM->getValue())->getSExtValue());
```
```suggestion
        cast<const ConstantInt>(CM->getValue())->getSExtValue());
```

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
+
+template <class Container>
+void CheckDeclaredAspects(LLVMContext &C, const Function *F,
+                          const Container &Aspects) {
```
```suggestion
                          const Container &UsedAspects) {
```

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
+    std::string Msg;
+    std::string AspectsStr = Join(MissedAspects, ',');
+    // TODO: demangle function name and aspect's IDs?
+    if (MissedAspects.size() > 1) {
+      Msg = formatv("for function \"{0}\" aspects [{1}] are missed in function "
+                    "declaration",
+                    F->getName(), AspectsStr);
+    } else {
+      Msg = formatv(
+          "for function \"{0}\" aspect {1} is missed in function declaration",
+          F->getName(), AspectsStr);
+    }
```
I suggest to unify this into something like "The following aspects are missing in function declaration: [list here]"

AlexeySachkov commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -196,6 +199,76 @@ template <typename T> void PrintMap(const T &m) {
+          F->getName(), AspectsStr);
+    }
+
+    C.diagnose(MissedAspectDiagnosticInfo(Msg));
```
```suggestion
    emitWarning(C, Msg);
```

?

maksimsab committed:
move PropagateAspectUsage Pass in SYCLLowerIR component

maksimsab committed:
refactor code

maksimsab commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,282 @@
+#include "llvm/SYCLPropagateAspectUsage/PropagateAspectUsage.h"
+#include "../IR/LLVMContextImpl.h"
```
I differentiate types by `llvm::Type*` rather than by string names. The only way I found to get pointers for types I was looking for (including DoubleTy) is `NamedStructTypes` in `LLVMContextImpl`. Unfortunately, I didn't find a way to do it in [LLVMContext](https://llvm.org/doxygen/classllvm_1_1LLVMContext.html|https://llvm.org/doxygen/classllvm_1_1LLVMContext.html).

AlexeySachkov replied:
For struct types see [my other comment](https://github.com/intel/llvm/pull/5348#discussion_r788820250), for `double` you can just use [`Type::getDoubleTy`](https://llvm.org/doxygen/classllvm_1_1Type.html#acb145f988329d1d621f73abcafea21d8)

maksimsab commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,282 @@
+  return new SYCLPropagateAspectUsageLegacyPass();
+}
+
+using TypesMap = std::unordered_map<std::string, SmallVector<int>>;
```
Leftover from first version of code which didn't work in result. Deleted it.

maksimsab commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,282 @@
+    if (it2 == CI.NamedStructTypes.end()) {
+      dbgs() << "Can't find type in NamedStructTypes. name: " << name << "\n";
+      continue;
+    }
```
Didn't know about getIdentifiedStructTypes(). Probably, it is better.

maksimsab commented on llvm/lib/SYCLPropagateAspectUsage/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,282 @@
+TypesMap2::const_iterator
+CheckTypeContainsOptionalType(const Type *T, const TypesMap2 &Types) {
+  dbgs() << "CheckType: " << *T << ", ptr: " << T << "\n";
+  auto it = Types.find(const_cast<Type *>(T));
```
I found out that for `std::unordered_map` doesn't accept constant keys if Key type is a pointer (if it isn't a pointer than it is alright). Here is an [example](https://ideone.com/u7VF4u).

Here is a chain of functions which do not modify any Module's content. So all arguments are constant. And only here I made a quick hack to proceed with solution.

AlexeySachkov replied:
Then I suggest to switch to some other container implementation (perhaps one of LLVM-provides ones?) to avoid doing `const_cast`s

maksimsab replied:
I figured out that the problem was that it is required for container's template parameter be constant to be able to access elements by constant arguments.

For example, if you have `std::vector<T *> v;` and you want to insert `const Type*` then you need `std::vector<const Type*>`. And LLVM containers aren't different in this aspect.

maksimsab committed:
support complex types graphs. add more tests

maksimsab committed:
apply clang-format

maksimsab committed:
refactor code

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+  TypesWithAspects[DoubleTy].insert(6);
+
+  Types.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes()) {
```
Looking at the [implementation] of `getIdentifiedStructTypes`, it passes `NamedOnly = true` argument to [`TypesFinder::run`](https://llvm.org/doxygen/classllvm_1_1TypeFinder.html#a76e63ec5dc3405d799590195281313e2), which means that we might miss some unnamed structures.

However, I don't know how otherwise find such unnamed structures and whether they can appear in our inputs at all.
We probably need to note this moment  and try to explore it a bit more

AlexeySachkov replied:
Ok, I see, thanks for checking that!

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+AspectsTy GetAspectsUsedByInstruction(Instruction &I, TypesWithAspectsTy &Types,
+                                      TypesWithAspectsTy &Cache) {
+  Type *ReturnType = I.getType();
+  SmallSet<int, 4> Aspects = GetAspectsFromType(ReturnType, Types, Cache);
+  SmallSet<int, 4> Result = Aspects;
```
```suggestion
  AspectsSetTy Result = GetAspectsFromType(ReturnType, Types, Cache);
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+    const ConstantAsMetadata *CM =
+        dyn_cast<const ConstantAsMetadata>(MDN->getOperand(i));
+    DeclaredAspects.insert(
+        dyn_cast<const ConstantInt>(CM->getValue())->getSExtValue());
```
```suggestion
        cast<const ConstantInt>(CM->getValue())->getSExtValue());
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+      continue;
+
+    LLVMContext &C = F->getContext();
+    CheckDeclaredAspects(C, F, Aspects);
```
```suggestion
    CheckUsedAndDeclaredAspects(C, F, Aspects);
```

I think that we primarily checking _used_ aspects, because declaration of an aspect alone can't trigger any diagnostics, but a _use_ of undeclared aspect is what triggers the diagnostic message

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+  }
+}
+
+void CheckAllDeclaredAspects(FunctionToAspectsMapTy &Map) {
```
```suggestion
void CheckUsedAndDeclaredAspects(FunctionToAspectsMapTy &Map) {
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+                      SmallPtrSet<Function *, 16> &Visited) {
+  SmallSet<int, 8> LocalAspects;
+  for (Function *Callee : CG[F]) {
+    if (Visited.contains(Callee)) {
+      Visited.insert(Callee);
```
What is the point of inserting `Callee` into `Visited` if it is already there? It looks like the condition should be inverted with `!`

maksimsab replied:
fixed.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+/// doesn't get into infite loop.
+AspectsTy GetAspectsFromType(Type *T, TypesWithAspectsTy &Types,
+                             TypesWithAspectsTy &Cache) {
+  auto it = Types.find(T);
```
```suggestion
  auto It = Types.find(T);
```

AlexeySachkov replied:
This is still not applied

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+    return;
+
+  AspectsTy DeclaredAspects;
+  for (unsigned i = 0; i != MDN->getNumOperands(); ++i) {
```
```suggestion
  for (unsigned I = 0; I != MDN->getNumOperands(); ++I) {
```

AlexeySachkov reviewed (changes requested):
Looks better! Overall I think that some comments around functions are missing, which would explain what function does and what its arguments mean.

Haven't looked into tests yet

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+}
+
+TypesWithAspectsTy
+GetTypesWithAspectsFromModule(Module &M, TypesWithAspectsTy &TypesWithAspects) {
```
There are different semantics. Argument map consist of known types from metadata information. While returned Map consist of types which we figured out during the search in this function.
The good point in current version is that it is stupid and straightforward because everything is done step by step in main run() function at the end of the file.

maksimsab committed:
reformat code

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+  }
+
+  std::unordered_map<Type *, std::vector<Type *>> Edges;
+  for (Type *T : Types) {
```
I hope a comment over the whole function is enough.

AlexeySachkov replied:
It looks good to me, but I will let others decide whether it is enough for understanding the code or not

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+  Cache[T] = {};
+  AspectsSetTy Result;
+
+  for (size_t i = 0; i != T->getNumContainedTypes(); ++i) {
+    Type *TT = T->getContainedType(i);
```
```suggestion
  for (Type *TT : T->subtypes()) {
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+
+  SmallVector<int> MissedAspects;
+  for (int aspect : Aspects) {
+    if (DeclaredAspects.count(aspect) == 0) {
+      MissedAspects.push_back(aspect);
```
```suggestion
    if (DeclaredAspects.count(Aspect) == 0) {
      MissedAspects.push_back(Aspect);
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+    LLVMContext &C = F->getContext();
+    std::vector<Metadata *> AspectsMetadata;
+    AspectsMetadata.reserve(Aspects.size());
+    for (int aspect : Aspects) {
+      AspectsMetadata.push_back(ConstantAsMetadata::get(
+          ConstantInt::getSigned(Type::getInt32Ty(C), aspect)));
```
```suggestion
    for (int Aspect : Aspects) {
      AspectsMetadata.push_back(ConstantAsMetadata::get(
          ConstantInt::getSigned(Type::getInt32Ty(C), Aspect)));
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+}
+
+void CheckAllDeclaredAspects(FunctionToAspectsMapTy &Map) {
+  for (auto &it : Map) {
+    Function *F = it.first;
+    auto &Aspects = it.second;
```
```suggestion
  for (auto &It : Map) {
    Function *F = It.first;
    auto &Aspects = It.second;
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
+
+void CreateAspectMetadataForFunctions(FunctionToAspectsMapTy &Map) {
+  for (auto &it : Map) {
+    Function *F = it.first;
+    auto &Aspects = it.second;
```
```suggestion
  for (auto &It : Map) {
    Function *F = It.first;
    auto &Aspects = It.second;
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+    if (CC != CallingConv::SPIR_FUNC && CC != CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
```
```suggestion
    if (CC == CallingConv::SPIR_KERNEL) {
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+    for (auto &I : instructions(F)) {
+      SmallSet<int, 4> Aspects =
+          GetAspectsUsedByInstruction(I, TypesWithAspects, Cache);
+      FunctionToAspects[&F].insert(Aspects.begin(), Aspects.end());
```
Can it be written as `FunctionToAspects[&F] = Aspects` here?

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+
+/// Retrieves from metadata (intel_types_that_use_aspects) types
+/// and aspects these types depend on.
+TypeToAspectsMapTy GetTypesThatUseAspectsFromMetadata(Module &M) {
```
```suggestion
TypeToAspectsMapTy GetTypesThatUseAspectsFromMetadata(const Module &M) {
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+TypeToAspectsMapTy
+GetTypesWithAspectsFromModule(Module &M, TypeToAspectsMapTy &TypesWithAspects) {
```
```suggestion
GetTypesWithAspectsFromModule(const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
```

psamolysov-intel commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+}
+
+FunctionToAspectsMapTy
+GetFunctionsAspects(Module &M, TypeToAspectsMapTy &TypesWithAspects) {
```
```suggestion
GetFunctionsAspects(const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
```

maksimsab replied:
It can't be done easily because it would lead to change `Function &` to `const Function &` in loop. And it would lead to change in `FunctionToAspectsMapTy` type (Key would be required to be `const Function *`).

psamolysov-intel reviewed (commented):
I'm not sure, what is the best style in our project, should we pass `Module` and other LLVM's stuff in a function by constant reference if the object will not be changed there or a reference is enough (`const Module &M`  vs `Module &M`)?

maksimsab committed:
reformat code

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,363 @@
+using AspectsTy = SmallSet<int, 4>;
+using TypesWithAspectsTy = std::unordered_map<Type *, AspectsTy>;
+
+TypesWithAspectsTy GetTypesWithAspectsFromMetadata(Module &M) {
```
Added some.

AlexeySachkov commented on llvm/test/SYCLLowerIR/PropagateAspectUsagePass/composite_types3.ll:
```diff
@@ -0,0 +1,103 @@
+;
+; Test checks structures with pointers.
+
+; By default only %D is optional.
+; %A and %B are optional because they contain %D.
+; %C contains a pointer on %A. Because of that there is a path
+; to %D which means that %C is optional as well.
```
```suggestion
; In this test, the only type which is explicitly marked as optional is %D
; Various test cases check if the pass is able to deal with various
; type hierarchies and correctly propagate aspects.
;
; %A and %B are optional because they contain %D.
; %C contains a pointer on %A. Because of that there is a path
; to %D which means that %C is optional as well.
```

maksimsab committed:
fix bug and commit corresponding test

AlexeySachkov commented on llvm/test/SYCLLowerIR/PropagateAspectUsagePass/composite_types4.ll:
```diff
@@ -0,0 +1,80 @@
+  ret void
+}
+
+;     -----> A1
+;     |      |
+;     |      v
+;    *|      B1
+;     |     / \
+;     |    v   v
+;     E1<--C1   D
+;     |    ^
+;    *|    |*
+;     v    |
+;     F1-->G1-->D1
+;       *
```
I suggest that we remove `D` from this test case and check aspects of type `A` to test aspect propagation through a more interesting path

AlexeySachkov commented on llvm/test/SYCLLowerIR/PropagateAspectUsagePass/empty.ll:
```diff
@@ -0,0 +1,20 @@
+; RUN: opt --PropagateAspectUsage < %s -S | FileCheck %s
```
I think that we should also add a test where a function uses two types that share the same aspect attached to them - to check that we don't emit duplicates in `!intel_used_aspects` metadata

maksimsab replied:
added one.

AlexeySachkov reviewed (commented):
A few comments about testing. Thanks for those ascii-pictures of graphs!

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,359 @@
+  while (!queue.empty()) {
+    Type *T = queue.front();
+    queue.pop();
+    if (Visited.count(T))
```
Consider the following example:
```
1->2
|  |
|  v
-->3
```
If we start from 1 then we push 2 and 3 into queue. Then we extract 2 and push 3 one more time since 3 is not present in Visited yet. Now we have two copies of vertex 3 in queue. For such cases the check in line 96 is introduced.

psamolysov-intel replied:
Hm... Thank you for the explanation.

maksimsab committed:
reformat code

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,386 @@
+
+    AspectsSetTy &Aspects = Result[T];
+    for (size_t I = 1; I != N.getNumOperands(); ++I) {
+      ConstantAsMetadata *CAM = dyn_cast<ConstantAsMetadata>(N.getOperand(I));
+      assert(CAM && "constant metadata is expected in "
+                    "intel_types_that_use_aspect list's entry");
```
Yes, but here we also have the message. Wouldn't it be helpful?

AlexeySachkov replied:
I think that the code around is pretty self-explanatory

AlexeySachkov replied:
The better point would be: `dyn_cast` performs a runtime check and can return `nullptr`. We do not check that `CAM` is `nullptr` at runtime and therefore we shouldn't be using `dyn_cast`. If we replace `dyn_cast` with `cast`, then that extra assertion won't be needed, because we will never reach it because there is a similar assertion within `cast`.

maksimsab committed:
add consts

maksimsab commented:
Added consts where it is possible. In some places it is hard because it would require to add const inside some using aliases which will break the code.

maksimsab committed:
edit tests

maksimsab committed:
make tests more reliable

maksimsab committed:
add test for unnamed struct

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+#include "llvm/SYCLLowerIR/PropagateAspectUsage.h"
```
Header with the license is missing

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+  static constexpr int AspectFP64 = 6; // TODO: add the link to spec
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
```
`DoubleTy` doesn't contain any other types, so it won't create any edges in types graph, why do we need to add it here as a type to process?

maksimsab replied:
This type corresponds to vertexes in a type graph. So there will be edges from DoubleTy to types which contain doubles.

maksimsab replied:
There is a piece of code below.
```cpp
if (!TypesToProcess.count(TT))
  continue;
```

Here is crucial to have DoubleTy in set.

AlexeySachkov replied:
Oh, got it, thanks

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+///
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+TypeToAspectsMapTy
+GetTypesWithAspectsFromModule(const Module &M,
+                              TypeToAspectsMapTy &TypesWithAspects) {
```
It still looks weird to me that in order to get types with aspects from a module, we need to pass types with aspects.
What about renaming it into something like `PropagateAspectsToOtherTypesInModule` ?

Ideally, we shouldn't be even returning any values, because we can directly modify `TypesWithAspects` map which is passed to us, but that will require preserving the map keys into a temporary store before iterating over map (to avoid possible issues with inserting during iterating)

maksimsab replied:
Changed it.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes()) {
+    TypesToProcess.insert(T);
+  }
```
```suggestion
  for (Type *T : M.getIdentifiedStructTypes())
    TypesToProcess.insert(T);
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+  for (const Type *T : TypesToProcess) {
+    for (const Type *TT : T->subtypes()) {
+      // If TT = %A*** then we want to get TT = %A
+      while (TT->isPointerTy()) {
+        TT = TT->getContainedType(0);
+      }
+
+      if (!TypesToProcess.count(TT)) {
+        continue; // We are not interested in some types. For example, IntTy.
+      }
```
```suggestion
      while (TT->isPointerTy())
        TT = TT->getContainedType(0);

      if (!TypesToProcess.count(TT))
        continue; // We are not interested in some types. For example, IntTy.
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+  }
+
+  TypeToAspectsMapTy Result;
+  for (const auto &It : TypesWithAspects) {
+    PropagateAspectsThroughTypes(Edges, It.first, It.second, Result);
+  }
```
```suggestion
  for (const auto &It : TypesWithAspects)
    PropagateAspectsThroughTypes(Edges, It.first, It.second, Result);
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+/// Checks that all declared function's aspects correspond to the
+/// aspects set @Aspects. If there is a inconsistency then corresponding
+/// warning is emitted.
+template <class Container>
```
BTW, do we really need this to be a function template?

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+  }
+
+  AspectsSetTy MissedAspects;
+  for (int Aspect : UsedAspects) {
+    if (DeclaredAspects.count(Aspect) == 0) {
+      MissedAspects.insert(Aspect);
+    }
+  }
```
```suggestion
  for (int Aspect : UsedAspects) {
    if (DeclaredAspects.count(Aspect) == 0)
      MissedAspects.insert(Aspect);
  }
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+    if (CC != CallingConv::SPIR_FUNC && CC != CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (CC == CallingConv::SPIR_KERNEL) {
+      Kernels.push_back(&F);
+    }
```
```suggestion
    if (CC == CallingConv::SPIR_KERNEL)
      Kernels.push_back(&F);
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+  }
+
+  SmallPtrSet<Function *, 16> Visited;
+  for (Function *F : Kernels) {
+    PropagateAspectsThroughCG(F, CG, FunctionToAspects, Visited);
+  }
```
```suggestion
  for (Function *F : Kernels)
    PropagateAspectsThroughCG(F, CG, FunctionToAspects, Visited);
```

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy GetAspectsFromType(const Type *T, TypeToAspectsMapTy &Types) {
+  auto it = Types.find(T);
+  if (it != Types.end()) {
+    return it->second;
+  }
```
```suggestion
  if (it != Types.end())
    return it->second;
```

AlexeySachkov reviewed (commented):
Overall LGTM to be promoted to a final review from a draft, mostly styling comments

maksimsab committed:
fix tests

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+      continue;
+    }
+
+    for (const Type *TT : Edges.at(T)) {
+      if (Visited.count(TT))
+        continue;
+
+      queue.push(TT);
+    }
```
```suggestion
    for (const Type *TT : Edges.at(T))
      if (!Visited.count(TT))
        queue.push(TT);
```

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+        TT = TT->getContainedType(0);
+      }
+
+      if (!TypesToProcess.count(TT)) {
+        continue; // We are not interested in some types. For example, IntTy.
+      }
+
+      Edges[TT].push_back(T);
```
```suggestion
      if (TypesToProcess.count(TT)) {
        // We are not interested in some types. For example, IntTy.
        Edges[TT].push_back(T);
      }
```

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,391 @@
+    return it->second;
+  }
+
+  // NB! This is essential to no get into infinite recursive loops.
```
```suggestion
  // NB! This is essential to avoid infinite recursive loops.
```

mlychkov replied:
I think it is not necessary to add "NB!" to comments.

maksimsab committed:
add test

maksimsab committed:
add comment for fp64 aspect

maksimsab committed:
reformat code

maksimsab committed:
add PropagateAspectUsage pass in clang

psamolysov-intel commented on llvm/include/llvm/LinkAllPasses.h:
```diff
@@ -207,6 +208,7 @@ namespace {
       (void) llvm::createExpandMemCmpPass();
       (void) llvm::createExpandVectorPredicationPass();
       (void)llvm::createESIMDVerifierPass();
```
I see you has not touched this line, but maybe it could be aligned with the line above and (new one) below? What about to add a space after `(void)`?

maksimsab replied:
In fact, I realized that clang-format prefers `(void)llvm::create();` without whitespace.

mlychkov commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -863,8 +864,10 @@ void EmitAssemblyHelper::CreatePasses(legacy::PassManager &MPM,
-  if (LangOpts.SYCLIsDevice)
+  if (LangOpts.SYCLIsDevice) {
     MPM.add(createESIMDVerifierPass());
+    MPM.add(createPropagateAspectUsagePass());
```
Could you please also include this pass into a pipeline for the new PM?

maksimsab replied:
I have a guess that it has already been added in pipeline since it was added in `PassRegistry.def` which is processed in `registerModuleAnalyses` method which is invoked from `RunOptimizationPipeline` method of `EmitAssemblyHelper`.
I've checked it with `-mllvm -debug-pass=Structure` and this pass appears in the list. I amn't sure that `RunOptimizationPipeline` is the best part since it has no common with actual optimizations. Maybe it would be better to turn it on separately in the main method in `EmitAsseblyHelper` like it is done in legacy part.

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes())
+    TypesToProcess.insert(T);
+
+  TypesEdgesTy Edges;
+  for (const Type *T : TypesToProcess) {
```
```suggestion
  TypesEdgesTy Edges;
  for (Type *T : M.getIdentifiedStructTypes())
```

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+  Types[T] = {};
+  AspectsSetTy Result;
+
+  for (const Type *TT : T->subtypes()) {
```
Why we collect aspects through tree again? Shouldn't Types[T] already contain all aspects for it and its subtypes after `PropagateAspectsThroughTypes`?

maksimsab replied:
`PropagateAspectsThroughTypes` tries to traverse over Identified types like structs mostly. While `getAspectsFromType` gets arguments from instructions which may operate with non-identified types. For example, there can be
`bitcast i64 %a to %T***`. Here, we want to get `%T` instead of `%T***`. The same with arrays and vectors.

mlychkov reviewed (changes requested):
My review is still in progress. First part of comments.

maksimsab committed:
add PropagateAspectUsagePass for new pass manager

maksimsab committed:
reformat code

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+void PropagateAspectsThroughTypes(const TypesEdgesTy &Edges, const Type *Start,
+                                  const AspectsSetTy &AspectsToPropagate,
+                                  TypeToAspectsMapTy &ResultAspects) {
+  SmallPtrSet<const Type *, 16> Visited;
+  std::queue<const Type *> queue;
```
It isn't exactly clear to me how to use it here.

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
```
But metadata contains nothing about double.

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+  TypesWithAspects[DoubleTy].insert(AspectFP64);
+
+  TypesToProcess.insert(DoubleTy);
+  for (Type *T : M.getIdentifiedStructTypes())
```
I think, they are handled in `getAspectsFromType` procedure.

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsUsedByInstruction(const Instruction &I,
+                                         TypeToAspectsMapTy &Types) {
+  Type *ReturnType = I.getType();
```
What if this type is a pointer?

mlychkov replied:
Ok, sorry, I thought that `subtypes()` returns types for structs only. But I've looked into `llvm::Type`'s implementation and now I see that all types are covered.

maksimsab replied:
`composite_types1.ll` checks that since alloca returns a pointer.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+}
+
+void checkUsedAndDeclaredAspects(const FunctionToAspectsMapTy &Map) {
+  for (const auto &It : Map) {
+    const Function *F = It.first;
+    auto &Aspects = It.second;
+    if (Aspects.empty())
+      continue;
+
+    LLVMContext &C = F->getContext();
+    checkDeclaredAspectsForFunction(C, F, Aspects);
+  }
```
+1 here, we can move querying `LLVMContext` down to `checkDeclaredAspectsForFunction` to make it signature a bit simpler

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+    std::string AspectsStr = join(MissedAspects, ',');
+    // TODO: demangle function name and aspect's IDs?
+    std::string Msg = formatv(
+        "for function \"{0}\" there is the list of missed aspects: [{1}]",
```
```suggestion
        "function \"{0}\" uses the following aspects without declaring them: [{1}]",
```

maksimsab replied:
Changed the format to the version from Design Doc.

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,393 @@
+  void print(DiagnosticPrinter &DP) const override { DP << Msg; }
+};
+
+void emitWarning(LLVMContext &C, const StringRef Msg) {
+  C.diagnose(MissedAspectDiagnosticInfo(Msg));
+}
```
This one liner is called only in one place. I think that it isn't worth defining a function for that.

maksimsab committed:
add support of call chains and debug information. add corresponding
tests

maksimsab committed:
fix remarks

maksimsab commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -237,28 +240,62 @@ void emitWarning(LLVMContext &C, const StringRef Msg) {
-      S += sep;
-      S += ' ';
-    }
+///
```
Add desc

maksimsab committed:
fix bug of extensions/inline_asm.cpp

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,491 @@
+      // If TT = %A*** then we want to get TT = %A
+      while (TT->isPointerTy())
+        TT = TT->getPointerElementType();
+
+      // We are not interested in some types. For example, IntTy.
+      if (!TypesToProcess.count(TT))
+        continue;
+
+      Edges[TT].push_back(T);
```
```suggestion
      // We are not interested in some types. For example, IntTy.
      if (TypesToProcess.count(TT))
        Edges[TT].push_back(T);
```

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,491 @@
+  }
+
+  TypeToAspectsMapTy Result;
+  for (const Type *T : TypesToProcess) {
+    propagateAspectsThroughTypes(Edges, T, TypesWithAspects);
+  }
```
```suggestion
  for (const Type *T : TypesToProcess)
    propagateAspectsThroughTypes(Edges, T, TypesWithAspects);
```

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,491 @@
+/// It encompases composite structures and pointers.
+/// NB! This function inserts new records in @Types map for new discovered
+/// types. For the best perfomance pass this map in the next invocations.
+AspectsSetTy getAspectsFromType(const Type *T, TypeToAspectsMapTy &Types) {
+  auto It = Types.find(T);
+  if (It != Types.end())
+    return It->second;
+
+  // This is essential to no get into infinite recursive loops.
+  Types[T] = {};
+  AspectsSetTy Result;
+
+  for (const Type *TT : T->subtypes()) {
+    AspectsSetTy Aspects = getAspectsFromType(TT, Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  Types[T] = Result;
+  return Result;
+}
```
```suggestion
const AspectsSetTy &getAspectsFromType(const Type *T, TypeToAspectsMapTy &Types) {
  auto It = Types.find(T);
  if (It != Types.end())
    return It->second;

  // This is essential to no get into infinite recursive loops.
  AspectsSetTy &Result = Types[T];

  for (const Type *TT : T->subtypes()) {
    const AspectsSetTy &Aspects = getAspectsFromType(TT, Types);
    Result.insert(Aspects.begin(), Aspects.end());
  }

  return Result;
}
```

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,491 @@
+  AspectsSetTy Result = getAspectsFromType(ReturnType, Types);
+  for (const auto &OperandIt : I.operands()) {
+    Type *T = OperandIt->getType();
+    AspectsSetTy Aspects = getAspectsFromType(T, Types);
```
```suggestion
    const AspectsSetTy &Aspects = getAspectsFromType(T, Types);
```

maksimsab committed:
Make warnings work with clang FE.
Add tests which check presence of PropagateAspectUsage pass in pipeline.
Add test which checks generated IR from C++ code.

againull commented on sycl/test/basic_tests/aspect_usage.cpp:
```diff
@@ -0,0 +1,50 @@
+// CHECK-WARNINGS-NEXT:  _Z5func3v()
+// CHECK-WARNINGS-NEXT: compile with '-g' to get source location
+
+// CHECK-WARNINGS-DBG: warning: function '_Z5func1v' uses aspect '5' not listed in `sycl::requires()`
```
Could you please clarify why warning message says "sycl::requires()"? Shouldn't the aspect be listed in "sycl::device_has()"? Is "sycl::requires()" is the same as "sycl::device_has()"?

maksimsab replied:
Yes, you're right. I copied it from [design doc](https://github.com/gmlueck/llvm/blob/1e927c107a7ac5c486a5620fba4ce80912701785/sycl/doc/OptionalDeviceFeatures.md#new-llvm-ir-pass-to-propagate-aspect-usage). In fact, it should be `sycl::device_has()`

maksimsab committed:
fix remarks

maksimsab committed:
remove warning tests from llvm tests. replace require with device_has

maksimsab committed:
add test cases for arrays and vectors

maksimsab committed:
fix aspect_usage.cpp test

mlychkov commented on clang/lib/CodeGen/BackendUtil.cpp:
```diff
@@ -1380,6 +1384,9 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
 
   ModulePassManager MPM;
 
+  if (CodeGenOpts.DisableLLVMPasses && LangOpts.SYCLIsDevice)
+    MPM.addPass(PropagateAspectUsagePass());
```
```suggestion
  if (LangOpts.SYCLIsDevice)
    MPM.addPass(PropagateAspectUsagePass());
```

AlexeySachkov replied:
`DisableLLVMPasses` check is needed here or otherwise we will add the pass twice. However, I would prefer if we only had a single insertion point for this pass instead of two like it is suggested in this PR

maksimsab replied:
In case `DisableLLVMPasses == false` we should add this pass in `registerPipelineStartEPCallback`, otherwise we should add just with `addPass` method. This detail is the reason why there are 2 insertion points.

mlychkov replied:
It is not necessary add this pass via `registerPipelineStartEPCallback` in general. I proposed such solution when there were no case for `DisableLLVMPasses == true`.
Taking into account a requirement to add this pass even for `DisableLLVMPasses == false` case it is better to add it once on line 1388.

mlychkov replied:
Ok, @maksimsab helped me to figure out that we should add `PropagateAspectUsagePass` in two places because later in `if (!CodeGenOpts.DisableLLVMPasses)` branch `ModulePassManager` instance is recreated and all that is added via `addPass` method before that moment is lost. On the other hand, functions which really add passes that are included via `registerPipeline*Callback` methods also called inside that branch.

mlychkov commented on llvm/include/llvm/IR/DiagnosticInfo.h:
```diff
@@ -1092,6 +1093,34 @@ class DiagnosticInfoDontCall : public DiagnosticInfo {
 };
 
+/// Class for emiting warning messages from SYCL pass - PropagateAspectUsage
+class DiagnosticInfoSYCLWarning : public DiagnosticInfo {
```
I think name of this diagnostic should be more specific if it is used only for PropagateAspectUsage pass. Like  DiagnosticInfoSYCLUnspecAspect.

AlexeySachkov replied:
Personally, I'm fine with the diagnostic class being named in generic way even if it is used in only one place. But, looking at the constructor arguments the diagnostic doesn't seem to be so generic, so we should either rename it to make it less generic or we should change the constructor (and other methods) to be less aspects-oriented to the class is indeed generic.

I'm not sure that we will need such kind of diagnostic any time soon, so renaming as @mlychkov suggests is probably preferable

AlexeySachkov commented on llvm/include/llvm/IR/DiagnosticInfo.h:
```diff
@@ -1092,6 +1093,34 @@ class DiagnosticInfoDontCall : public DiagnosticInfo {
+  SmallString<16> FunctionName;
+  SmallString<5> AspectStr;
+  std::string CallChain;
+  bool IsFullDebugMode = 0;
```
```suggestion
  bool IsFullDebugMode = false;
```

AlexeySachkov commented on llvm/lib/IR/DiagnosticInfo.cpp:
```diff
@@ -431,3 +431,5 @@ void DiagnosticInfoDontCall::print(DiagnosticPrinter &DP) const {
     DP << ": " << getNote();
 }
+
+void DiagnosticInfoSYCLWarning::print(DiagnosticPrinter &DP) const {}
```
We should at least have a `FIXME: implement` here if we do not plan to implement this method immediately

maksimsab replied:
I am not sure that this method should be implemented

mlychkov commented:
What about `!intel_fixed_targets` metadata? Will it be added here or in a separate patch?

AlexeySachkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,462 @@
+        if (It == AspectsWithFunctionLinkSet.end())
+          continue;
+
+        AspectWithFunctionLink = *It;
+        AspectsWithFunctionLinkSet.erase(AspectWithFunctionLink);
+        AspectWithFunctionLink.DL = &I.getDebugLoc();
+        AspectsWithFunctionLinkSet.insert(AspectWithFunctionLink);
```
Why do we need `erase` + `insert`, why can't we just update `.DL` field?

maksimsab replied:
`AspectWithFunctionLinkSet` is a SmallSet. No set-like container allows to modify its element since order might be broken. However, I can change SmallSet to some map container.

AlexeySachkov commented on sycl/test/basic_tests/aspect_usage.cpp:
```diff
@@ -0,0 +1,50 @@
+// RUN: %clangxx -fsycl -fno-legacy-pass-manager %s -g -emit-llvm -S -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS-DBG -DPATH=%s
+// RUN: %clangxx -fsycl -flegacy-pass-manager %s -g -emit-llvm -S -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS-DBG -DPATH=%s
+
+// CHECK-WARNINGS: warning: function '_Z5func1v' uses aspect '5' not listed in `sycl::device_has()`
+// CHECK-WARNINGS-NEXT:  _Z5func1v()
```
We should have some clarification words added to the diagnostic or otherwise it might not be clear what are those function names below. Something like `Call stack:` perhaps?

maksimsab replied:
There is a line. There was no check for it but I will add it.

AlexeySachkov commented on sycl/test/basic_tests/aspect_usage.cpp:
```diff
@@ -0,0 +1,50 @@
+
+// CHECK-WARNINGS: warning: function '_Z5func1v' uses aspect '5' not listed in `sycl::device_has()`
+// CHECK-WARNINGS-NEXT:  _Z5func1v()
+// CHECK-WARNINGS-NEXT:  _Z5func2v()
```
Why do we need those `()` - will they be filled with the actual function prototype if it has arguments?

maksimsab replied:
Currently, arguments info is not emitted. Theoretically, it is easy with scalar arguments since I have function's signature in IR. However, I do not have a clear vision how to handle template arguments like `int func(MyStruct<T> &MS)`.

maksimsab replied:
In fact, I've realized that I've already added a demangling some time ago but haven't updated test. Now test contains arguments in functions.

AlexeySachkov commented on sycl/test/basic_tests/aspect_usage.cpp:
```diff
@@ -0,0 +1,50 @@
+// CHECK-WARNINGS-NEXT: compile with '-g' to get source location
+
+// CHECK-WARNINGS-DBG: warning: function '_Z5func1v' uses aspect '5' not listed in `sycl::device_has()`
+// CHECK-WARNINGS-DBG-NEXT:  _Z5func1v() [[PATH]]:45:57
+// CHECK-WARNINGS-DBG-NEXT:  _Z5func2v() [[PATH]]:43:22
+// CHECK-WARNINGS-DBG-NEXT:  _Z5func3v() [[PATH]]:38:10
```
Can we improve the call stack message here? I mean that it would be great to give user more info about what those names mean, something like:

```
in function _Z5func1v (defined at [[PATH]]:45:57)
which calls _Z5func2v (defined at [[PATH]]:43:22)
which calls _Z5func3v (defined at [[PATH]]:38:10)
which uses aspect '5'
```

Or otherwise:
```
aspect 5 is used in function _Z5func3v (defined at [[PATH]]:38:10)
which is called by _Z5func2v (defined at [[PATH]]:43:22)
which is called by _Z5func1v (defined at [[PATH]]:45:57)
```

AlexeySachkov commented:
> What about !intel_fixed_targets metadata? Will it be added here or in a separate patch?

This patch is already pretty big. I suggest we do not add new functionality in here and postpone it to separate PRs

mlychkov commented:
> > What about !intel_fixed_targets metadata? Will it be added here or in a separate patch?
> 
> This patch is already pretty big. I suggest we do not add new functionality in here and postpone it to separate PRs

Ok, I have no objection to a separate patch, but it would be nice then to add a remark to a commit message that some functionality will be added later.

maksimsab committed:
add  aspect id demangling

maksimsab committed:
fix failing tests

maksimsab committed:
fix remarks

maksimsab committed:
add test for composite types with vectors. fix aspect_usage.cpp test

maksimsab committed:
fix compile error

maksimsab committed:
Merge branch 'sycl' of https://github.com/intel/llvm into new_pass

maksimsab committed:
fix compile error

maksimsab committed:
fix aspect_usage.cpp test

maksimsab committed:
add demangling for function name

maksimsab committed:
unify pass names from both PMs

maksimsab commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -123,6 +123,7 @@ MODULE_PASS("LowerESIMD", SYCLLowerESIMDPass())
 MODULE_PASS("ESIMDLowerVecArg", ESIMDLowerVecArgPass())
 MODULE_PASS("esimd-verifier", ESIMDVerifierPass())
 MODULE_PASS("SYCLMutatePrintfAddrspace", SYCLMutatePrintfAddrspacePass())
+MODULE_PASS("SYCLPropagateAspectUsage", PropagateAspectUsagePass())
```
changed to sycl-propagate-aspect-usage.

AlexeySachkov reviewed (dismissed):
Overall, LGTM. I suggest that we merge this change and if there would be further comments, we apply them post-commit.

mlychkov reviewed (dismissed):
Ok, I have no more major comments now. Thank you for the patch.

mlychkov commented:
@maksimsab There is a failing test in pre-commit. Could you please check it?

AlexeySachkov committed:
Fix clang-format

Fznamznon commented on clang/lib/CodeGen/CodeGenAction.cpp:
```diff
@@ -825,6 +826,14 @@ void BackendConsumer::DontCallDiagHandler(const DiagnosticInfoDontCall &D) {
 
+void BackendConsumer::SYCLWarningDiagHandler(
+    const DiagnosticInfoSYCLUnspecAspect &D) {
+  SourceLocation Loc;
+  Diags.Report(Loc, diag::warn_sycl_wrong_aspect_usage)
```
Oh, it uses invalid source location, so the warning won't point to real code. Maybe we can try to construct valid source info at least when we have debug info? Seems like `DontCallDiagHandler`  defined above at least tries something like that.

AlexeySachkov replied:
This is a good idea and it seems like that is a better approach than we have right now. However, `DontCallDiagHandler` relies on source location emitted into LLVM IR by FE in form of `srcloc` metadata: [D106030](https://reviews.llvm.org/D106030), [D110364](https://reviews.llvm.org/D110364).

Amount of the changes in this PR is already huge, would it be acceptable to postpone this diagnostic improvement till the next PR?

Fznamznon replied:
> Amount of the changes in this PR is already huge, would it be acceptable to postpone this diagnostic improvement till the next PR?

I think it is ok to postpone. Please file a tracker if you're not going to work on it ASAP.

AlexeySachkov replied:
> I'm less certain it's okay to postpone as this means all diagnostics will show up in very wrong places (if at all) within an IDE, for example. I don't think the user experience is all that great when the diagnostic is a warning that's divorced from their code.

Note that we are in the very beginning of our optional kernel features implementation journey and even if user properly sets all the attributes, the code won't work as expected, because we are missing a lot of changes in other components. Therefore, I think that it should be ok not to complicate this PR even more, but still refactor diagnostics _before_ we mark the whole feature as completed.

> Note: losing source location information and providing a poor user experience with diagnostics is why very few diagnostics ever come from the backend. It's too easy to lose the source locations, especially when optimizer passes get involved. That's why the majority of backend diagnostics are in the form of remarks instead of warnings (let alone errors). We should be very careful about the user experience here.

That pass is supposed to be launched before all other passes or otherwise not only source locations will be lost, but also some semantics, i.e. we can lose information about declared aspects due to inlining.

Fznamznon replied:
> Therefore, I think that it should be ok not to complicate this PR even more, but still refactor diagnostics before we mark the whole feature as completed.

I'm ok with this approach.

AlexeySachkov replied:
I've created #5877 to track this further improvement

Fznamznon commented on sycl/test/basic_tests/aspect_usage.cpp:
```diff
@@ -0,0 +1,50 @@
+// Test checks Propagate Aspect Usage pass.
```
It test warnings emission. I would expect test for warnings in clang tests.

AlexeySachkov replied:
Which directory would be the best for such test? I assume `SemaSYCL` is not applicable here, because no changes to Sema were introduced. Perhaps `CodeGenSYCL`?

elizabethandrews replied:
Diagnostic tests are usually added to SemaSYCL

AlexeySachkov replied:
> Diagnostic tests are usually added to SemaSYCL

Because those diagnostics are emitted by `Sema`. I'm just not sure that test for a non-`Sema` component should be in `Sema`: we can't even use `-fsyntax-only` in that test. Diagnostics here are essentially emitted from post-`CodeGen`, so to me `CodeGen` seems like the best fit

Fznamznon replied:
https://reviews.llvm.org/D106030 adds test to clang/test/Frontend directory. I suggest to be consistent with community code and do the same.

AlexeySachkov replied:
Added `Frontend` test in 75d53aee043a341cb9448ecdafde583876718efb

bader replied:
Perhaps @AaronBallman can answer this as he reviewed https://reviews.llvm.org/D106030. I don't know what is the right directory for the tests checking diagnostics emitted from LLVM passes.

AaronBallman replied:
Tests in Sema should almost always be `-fsyntax-only` which means we never even emit the LLVM IR let alone run the backend. I think Frontend is a good place for the diagnostic tests as that's what eventually reports them out to the user. I would not make a FrontendSYCL directory as I don't expect us to have a significant number of tests in it. If we find that we do need quite a few tests (a dozen-ish?) then it may make sense to add the directory at that time.

Fznamznon reviewed (commented):
I've never seen diagnostics that come from LLVM pass, but emitted by clang. Interesting. Although I have concerns about how they will look like for the end users. 
I think we have some passes that emit errors, but they don't involve clang, so the quality of diagnostics is the same (no source info advantage), but the code is simpler. See https://github.com/intel/llvm/pull/5706 for example. Maybe it will be better?

elizabethandrews commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -313,4 +313,11 @@ def warn_profile_data_unprofiled : Warning<
 } // end of instrumentation issue category
 
+def warn_sycl_wrong_aspect_usage : Warning<
+  "function '%0' uses aspect '%1' not listed in `sycl::device_has()`\n"
```
> I've never seen diagnostics that come from LLVM pass, but emitted by clang. Interesting. Although I have concerns about how they will look like for the end users. I think we have some passes that emit errors, but they don't involve clang, so the quality of diagnostics is the same (no source info advantage), but the code is simpler. See #5706 for example. Maybe it will be better?

Yea I'm not familiar with this either. @maksimsab are there some other examples you can point us to? I'm not sure how to review FE changes here because I've never seen it done this way.

AlexeySachkov replied:
Diagnostic message was refactored in f381619913908ff90da025bf1e50be4ef76acd26

elizabethandrews commented:
> I've never seen diagnostics that come from LLVM pass, but emitted by clang. Interesting. Although I have concerns about how they will look like for the end users. I think we have some passes that emit errors, but they don't involve clang, so the quality of diagnostics is the same (no source info advantage), but the code is simpler. See #5706 for example. Maybe it will be better?

Yea I'm not familiar with this either. @maksimsab are there some other examples you can point us to? I'm not sure how to review FE changes here because I've never seen it done this way.

kbobrovs commented on llvm/test/SYCLLowerIR/PropagateAspectUsagePass/composite_types1.ll:
```diff
@@ -0,0 +1,93 @@
+  ret void
+}
+
+!intel_types_that_use_aspects = !{!0}
```
It would be great to have a handful of test cases checking that ESIMD kernels get proper aspects too.
Example SYCL ESIMD kernel and function for testing purposes:
```
define dso_local spir_kernel void @ESIMD_kernel() #0 !sycl_explicit_simd {
entry:
  ret void
}

define dso_local spir_func void @ESIMD_func() #0 !sycl_explicit_simd {
entry:
  ret void
}

attributes #0 = { "sycl-module-id"="a.cpp" }
```

AlexeySachkov replied:
I can add such test, but I don't quite understand why we need a handful of them - aspects propagation works regardless of ESIMD presence. 

If we want to ensure that particular high-level ESIMD constructs generate correct aspects in LLVM IR - then such tests should probably be added into `sycl/` sub-project together with the patch which assigns aspects to ESIMD constructs and types

kbobrovs replied:
I believe this patch should work equally well for SYCL and ESIMD kernels/functions, so not sure why ESIMD separation is needed here. At least two tests - kernel and function would be appreciated.

maksimsab replied:
Added such tests.

kbobrovs commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,527 @@
+    if (CC != CallingConv::SPIR_FUNC && CC != CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (CC == CallingConv::SPIR_KERNEL)
```
Module "roots" (or "entry points") for the aspect propagation should also include SYCL_EXTERNAL functions. In fact, there can be modules w/o kernels at all.
See [example ](https://github.com/intel/llvm/blob/4072557b121d0fb322a5e167cc8905209f21ab3e/llvm/tools/sycl-post-link/sycl-post-link.cpp#L304) from post-link how entry points can be determined.

AlexeySachkov replied:
Fixed and added a test in b558a2d1f3178d14ab74658750d00d8d1137dddb

AlexeySachkov commented:
> I think we have some passes that emit errors, but they don't involve clang, so the quality of diagnostics is the same (no source info advantage), but the code is simpler. See https://github.com/intel/llvm/pull/5706 for example. Maybe it will be better?

According to our implementation design document, we have settled on a warning diagnostic message and not an error. Therefore, we can't use `LLVMContext::emitError` method which would indeed simplify the implementation

Fznamznon commented:
> > I think we have some passes that emit errors, but they don't involve clang, so the quality of diagnostics is the same (no source info advantage), but the code is simpler. See #5706 for example. Maybe it will be better?
> 
> According to our implementation design document, we have settled on a warning diagnostic message and not an error. Therefore, we can't use `LLVMContext::emitError` method which would indeed simplify the implementation

Right. I didn't pay attention to the fact that you need a warning when I was typing this comment. Please ignore it.

AlexeySachkov committed:
Improve diagnostic message

AlexeySachkov committed:
Outline aspects enum into separate file to increase readability

AlexeySachkov committed:
Propagate aspects up to SYCL_EXTERNAL functions also

AlexeySachkov commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -313,4 +313,11 @@ def warn_profile_data_unprofiled : Warning<
+def warn_sycl_wrong_aspect_usage : Warning<
+  "function '%0' uses aspect '%1' not listed in `sycl::device_has()`\n"
+  "use is from this call chain:\n"
+  "%2"
```
Diagnostic message was improved in f381619913908ff90da025bf1e50be4ef76acd26

AaronBallman reviewed (commented):
Some comments from the peanut gallery.

AlexeySachkov committed:
Apply comments

AlexeySachkov committed:
Add Frontend test

AlexeySachkov commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -313,4 +313,11 @@ def warn_profile_data_unprofiled : Warning<
 
+def warn_sycl_wrong_aspect_usage : Warning<
+  "function '%0' uses aspect '%1' not listed in `sycl::device_has()`">,
+  InGroup<SyclAspectUsage>;
```
Applied in 82e4e1b128a1dca8196901b7eb5417ccc4e4184b

AlexeySachkov commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -1242,6 +1242,7 @@ def Sycl2020Compat : DiagGroup<"sycl-2020-compat">;
 def SyclStrict : DiagGroup<"sycl-strict", [ Sycl2017Compat, Sycl2020Compat]>;
 def SyclTarget : DiagGroup<"sycl-target">;
 def SyclFPGAMismatch : DiagGroup<"sycl-fpga-mismatch">;
+def SyclAspectUsage : DiagGroup<"sycl-aspect-usage">;
```
Removed in 82e4e1b128a1dca8196901b7eb5417ccc4e4184b

maksimsab commented:
> Yea I'm not familiar with this either. @maksimsab are there some other examples you can point us to? I'm not sure how to review FE changes here because I've never seen it done this way.

I was inspired by LTODiagnosticInfo a little bit. Of course, it is not common to emit warnings from backend.

maksimsab committed:
Merge branch 'sycl' of https://github.com/intel/llvm into HEAD

kbobrovs commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,492 @@
+  if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+    return true;
+
+  // FIXME: sycl-post-link allows t disable treating SYCL_EXTERNAL's as entry
```
```suggestion
  // FIXME: sycl-post-link allows to disable treating SYCL_EXTERNAL's as entry
```

maksimsab committed:
add test checking ESIMD metadata presence

maksimsab committed:
fix linter remark

mlychkov commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,492 @@
+    for (const Type *TT : T->subtypes()) {
+      // If TT = %A*** then we want to get TT = %A
+      // The same with arrays and vectors
+      while (TT->isPointerTy() || TT->isArrayTy() || TT->isVectorTy()) {
```
Typed pointers should be replaced by opaque pointers soon.
Could you please check that this code works for opaque pointers as well?

AlexeySachkov replied:
I don't see that any of those APIs is going to be removed and therefore this code should continue to compile even with opaque pointers and it will continue to work correctly for arrays and vectors.

Pointer "branch" of this code won't work as it used to, because I assume there would be no contained type in it anymore.

@maksimsab, I think we should do the following:
```suggestion
      while ((TT->isPointerTy() && !TT->isOpaquePointerTy()) || TT->isArrayTy() || TT->isVectorTy()) {
```

This will skip opaque pointers and they will be handled elsewhere when they are accessed and this code will work for non-opaque pointers as intended by looking through them.

maksimsab replied:
In fact, I realized that opaque pointers might break the current solution because the first step of algorithm (which traverses over type graph) becomes insufficient for overall correctness.

maksimsab committed:
allign aspect_usage.cpp test

elizabethandrews commented on clang/test/CodeGenSYCL/uses_aspects.cpp:
```diff
@@ -28,7 +28,7 @@ template <cl::sycl::aspect Aspect>
 
 [[__sycl_detail__::__uses_aspects__(cl::sycl::aspect::cpu)]] void func6();
-// CHECK: define dso_local spir_func void @{{.*}}func6{{.*}} !intel_used_aspects ![[ASPECTS1]] {
+// CHECK: define dso_local spir_func void @{{.*}}func6{{.*}} !intel_used_aspects ![[ASPECTS4:[0-9]+]] {
```
Why did this change? `func6 `should not have 2 aspects in IR as in your change

AlexeySachkov replied:
`func6` definition can be found down below and the function uses `Type2WithAspect`, which annotated with both `fp16` and `cpu` aspects

elizabethandrews commented on clang/test/CodeGenSYCL/uses_aspects.cpp:
```diff
@@ -57,10 +57,11 @@ void foo() {
+// CHECK-DAG: [[ASPECTS1]] = !{i32 1}
+// CHECK-DAG: [[ASPECTS2]] = !{i32 5, i32 2}
+// CHECK-DAG: [[ASPECTS3]] = !{i32 0}
+// CHECK-DAG: [[ASPECTS4]] = !{i32 1, i32 5}
```
This does not seem correct. Aspect 5 corresponds to fp16. Why is this generated for `func6`

AlexeySachkov replied:
See my comment above, `func6` uses `Type2WithAspect`, which is annotated with `fp16`

elizabethandrews commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -0,0 +1,42 @@
+// Test checks Propagate Aspect Usage pass.
+
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm %s -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-WARNINGS
```
Can we not use `-verify `here?

AlexeySachkov replied:
Unfortunately at the moment those diagnostics are emitted without source location attached to them and therefore I expect that `-verify` won't work. I've submitted #5877 to track further improvement of that diagnostic which allow us to use `-verify`

Fznamznon commented on clang/lib/CodeGen/CodeGenAction.cpp:
```diff
@@ -825,6 +826,27 @@ void BackendConsumer::DontCallDiagHandler(const DiagnosticInfoDontCall &D) {
+      << D.getFunctionName() << D.getAspect();
+
+  if (!D.getCallChain().empty()) {
+    Diags.Report(Loc, diag::note_sycl_wrong_aspect_usage)
+        << /* the actual use */ 0 << D.getCallChain().back()
+        << !D.isFullDebugMode();
+
+    for (auto I = ++D.getCallChain().rbegin(), E = D.getCallChain().rend();
+         I != E; ++I) {
+      Diags.Report(Loc, diag::note_sycl_wrong_aspect_usage)
```
So, do I understand it correctly that this will also be applied in the next PR?

bader commented on llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp:
```diff
@@ -0,0 +1,492 @@
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
```
Do I understand it correctly that here we are injecting dependencies on external project to the compiler?
Imagine how hard it will be to debug bugs in this code if I update **only** `sycl/include/CL/sycl/aspects.hpp`.

AlexeySachkov replied:
That is correct, both the compiler and the runtime must be in sync.

I don't really want to introduce a hard dependency between them and therefore I don't know what would be a good way to handle that. I suppose we should at least add a comment to `aspects.hpp` saying that if this file is changed, then `PorpagateAspectUsage` pass must also be changed, what do you think?

AlexeySachkov replied:
We had an offline sync with @bader and agreed that this such dependency on runtime headers is a design issue. We decided not to block the PR and I've created #5892 to track the problem

AlexeySachkov commented on llvm/lib/SYCLLowerIR/Aspects.def:
```diff
@@ -0,0 +1,38 @@
+#error "ASPECT macro is expected to be defined"
+#endif
+
+ASPECT(0, "host")
```
My understanding is that this file should be aligned with our implementation in headers to match values assigned there - from that point of view, I would prefer to follow the same order as in headers simply not to miss anything.

Whether or not that order in headers should be changed, why it contains `host` and whether or not it is ok to re-order  enum values (comparing to the spec) - those are questions to @intel/llvm-reviewers-runtime

https://github.com/intel/llvm/blob/c818709fcebd5d8183e0ebe127e7f37b8867c3f2/sycl/include/CL/sycl/aspects.hpp#L15-L17

AlexeySachkov replied:
We had an offline sync with @bader and agreed that this such dependency on runtime headers is a design issue. We decided not to block the PR and I've created #5892 to track the problem

AlexeySachkov replied:
@maksimsab, but we still need to (at least temporarily) add a comment into `sycl/include/CL/sycl/aspects.hpp` saying that whenever this file is changed, the pass (`Aspects.def`) must be adjusted as well

maksimsab committed:
fix type

Co-authored-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

maksimsab commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -0,0 +1,42 @@
+  int a = 0;
+};
+
+int func3(int a, int b, int c) {
```
Just to make functions different from each other. I can simplify them.

elizabethandrews replied:
Not required unless you want to. I'm ok with this.

maksimsab committed:
fix Frontend/sycl-aspects-usage.cpp test

maksimsab committed:
add considering of opaque pointers

maksimsab committed:
add comment to sycl/include/CL/sycl/aspects.hpp

maksimsab committed:
add few cases in sycl-aspects-usage test

maksimsab committed:
add check with several aspects in sycl-aspects-usage test

maksimsab commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -0,0 +1,42 @@
+
+class KernelName;
+
+struct [[__sycl_detail__::__uses_aspects__(aspect::fp16)]] Struct {
```
Added check of empty `device_has` and `double` type. Unfortunately, realized that `__uses_aspects__()` for function's declaration is not implemented yet. I think it would be better to implement it in the following patch.

maksimsab replied:
I am sorry. I meant that `__uses_aspects__()` is not handled in the proposed pass at the current moment. On the FE side it is implemented of course.

maksimsab committed:
clafiry sycl-aspects-usage test

maksimsab commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -0,0 +1,68 @@
+}
+
+[[sycl::device_has()]] int checkEmptyDeviceHas() {
+  return 0;
```
So if function `F1` with a set of declared aspects `A` calls function `F2` with a set of actually used aspects `B`. And `B` might not be inside of `A`. Let's `C` call a set `B-A`. Currently, only one arbitrary element from `C` would be printed. This is not user-friendly since a user would need to run compilation several times to figure out the whole set `C`.
We could print one call chain for every missed aspect from `C`. On the other hand, the size of output might blow up and the output will contain lots of repetitions and it might be difficult to extract additional information with comparison to the current way when we print a call chain for only one member of set `C`. Although I am little bit skeptical about printing everything, I am open to objections.

maksimsab replied:
I apologize. Checked one more time and currently for every missed aspect from `C` would be printed corresponding call chain.

elizabethandrews replied:
Sorry for not being clear. I was trying to say we need to test this case - 

```
[[sycl::device_has()]] int checkEmptyDeviceHas() {
   funcWithStruct();
}
```

Diagnostic should be generated since we have an empty aspect list.

We also need to test this case - 

```
[[sycl::device_has(aspect::fp16)]] int checkStructUsesAspect2( ) {
  funcWithStruct();
}

[[sycl::device_has()]] int checkEmptyDeviceHas() {
   checkStructUsesAspect2();
}
```

It is not clear to me what the required behavior is here. `checkStructUsesAspect2 `has attribute with `aspect::16`, but it's parent caller has empty aspect list. What shoudl requried behavior be here @gmlueck ?

gmlueck replied:
> ```
> [[sycl::device_has()]] int checkEmptyDeviceHas() {
>    funcWithStruct();
> }
> ```
> 
> Diagnostic should be generated since we have an empty aspect list.

Just to make sure we are on the same page, it is perfectly OK to have an empty aspect list in `[[sycl::device_has()]]`.  This means "my function doesn't require a device that has any specific aspects".  If such a function *does* use a feature that requires an aspect, then the compiler  should raise a diagnostic.  In your example above, does `funcWithStruct()` use a feature that requires an aspect?  Is this why you say that a diagnostic should be generated?


> We also need to test this case -
> 
> ```
> [[sycl::device_has(aspect::fp16)]] int checkStructUsesAspect2( ) {
>   funcWithStruct();
> }
> 
> [[sycl::device_has()]] int checkEmptyDeviceHas() {
>    checkStructUsesAspect2();
> }
> ```
> 
> It is not clear to me what the required behavior is here. `checkStructUsesAspect2 `has attribute with `aspect::16`, but it's parent caller has empty aspect list. What shoudl requried behavior be here @gmlueck ?

This should cause a diagnostic.  The `[[sycl::device_has(aspect::fp16)]]` on `checkStructUsesAspect2` is an assertion that this function requires that aspect.  See this sentence in SYCL 2020:

>When this attribute appears in a translation unit that calls the decorated device function, it is an assertion that the device function uses optional features that correspond to the aspects listed in the attribute.

The `[[sycl::device_has()]]` on `checkEmptyDeviceHas` asks the compiler to generate a diagnostic if that function uses any feature that requires any aspect (because the aspect list is empty).  Since `checkStructUsesAspect2` requires `fp16`, a diagnostic should be generated.

elizabethandrews replied:
>does funcWithStruct() use a feature that requires an aspect? Is this why you say that a diagnostic should be generated?

Yes `funcWithStruct()` uses a type marked with `aspect::fp16`

Thanks for the explanation. That makes sense. @maksimsab could you please add these tests to ensure behavior is as expected?

maksimsab replied:
Fixed `checkEmptyDeviceHas()` and added `checkEmptyDeviceHas2()` with your new example.

maksimsab committed:
fix checkSeveralAspects test

maksimsab commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -0,0 +1,68 @@
+  return static_cast<double>(s.a);
+}
+
+// Check that a warning diagnostic works in a case
```
I apologize. Added them.

elizabethandrews commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -20,12 +22,18 @@ struct [[__sycl_detail__::__uses_aspects__(aspect::fp16)]] Struct {
   return s.a;
 }
 
+[[sycl::device_has(aspect::fp64)]] int checkStructUsesAspect(int a) {
```
Unused function argument 'a' and for `funcWithStruct`

elizabethandrews commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -20,12 +22,18 @@ struct [[__sycl_detail__::__uses_aspects__(aspect::fp16)]] Struct {
+}
+
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm %s -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-EMPTY
+// CHECK-EMPTY-NOT: checkEmptyDiveceHas
```
I'm not sure what you are testing here. Can you please explain? Are you trying to test function is not generated?

maksimsab replied:
I was testing here that this function is not present in warning messages. I will fix it right now.

elizabethandrews replied:
Oh I see. You were testing warning was not being generated for empty argument list. I don't think you need to test that since that comes under semantic tests of `device_has` attribute which has already been tested in FE. 

I think in this PR you just need to test propagation and usage of aspects. So for a function with `device_has` attribute with empty argument list, a good test will be to use a type with aspect in that function to ensure warning is generated

elizabethandrews commented on clang/test/Frontend/sycl-aspects-usage.cpp:
```diff
@@ -0,0 +1,93 @@
+// with declared in device_has().
+
+// Check for fp64 aspect
+// RUN: %clang_cc1 -fsycl-is-device -triple spir64-unknown-unknown -emit-llvm %s -o /dev/null 2>&1 | FileCheck %s --check-prefix CHECK-SEVERAL-FP64
```
Actually, why do you have multiple run lines? All these warnings will be generated with the first run line in this test right?

maksimsab replied:
Basically, I do it that way to avoid issues with undefined ordering of resulted warnings.

maksimsab committed:
add checkEmptyDeviceHas2 test in sycl-aspects-usage test

elizabethandrews reviewed (approved):
FE changes LGTM for now. As discussed in PR, diagnostics should be fixed in a followup PR - https://github.com/intel/llvm/issues/5877
